### ISSUE TITLE
Decide on the Deployment Model

### ISSUE THREAD

**Issue by arkodg** [envoyproxy/gateway#97]
Outlining various deployment models that allow the project to achieve its [goals](https://github.com/envoyproxy/gateway/blob/main/GOALS.md)

*CP = Control Plane* 
*DP = Data Plane*

![EG Deployment Tradeoffs](https://user-images.githubusercontent.com/9469142/172480771-16ec3554-443c-432a-9399-22989d7ef112.png)

1. Combine CP and Envoy (DP) into a single deployment
* Similar to what is done in [Emissary ](https://github.com/emissary-ingress/emissary) today
* Every CP watches for a specific GatewayClass, as well as the Gateway and HTTPRoute resources linked to it
* Pros
   * Simple to deploy & operate
   * Makes internal & user cert management simpler (Unix domain sockets b/w Envoy and xDS Server)
   * Simpler upgrades
* Cons
  * More load on k8s API Server
  * Resource consumption increases - CPs are duplicating work
  * Aggregation of stats/status across all identical controllers moves into the CLI, external layer
  
2. Provision a unique CP and Envoy (DP) per GatewayClass
* Similar to what is done in [Contour](https://github.com/projectcontour/contour) today
* The main provisioner watches for a specific GatewayClass, and creates a CP per GatewayClass
* Each CP watches for Gateway and HTTPRoutes and configures the xDS Server tied to it
* Pros
   * Simpler CP
   * Envoy scales with App load but CP doesnt
   * API server load independent from App load
   * Isolates user certs by isolating xDS Servers
* Cons
  * Non trivial Authn + Authz between CP and Envoy
  * More containers/deployments to manage & operate
  * Another communication channel over the network between the main provisioner and every CP to aggregate stats/status
  * Non trivial upgrades
  
3. Common CP with a unique Envoy (DP) per GatewayClass
  * Similar to what is done in [Istio](https://github.com/istio/istio) today
  * The xDS Server is shared across all Envoys
  * The DP component also contains an agent (similar to Istio-agent) that performs SDS local to the Envoy
  
 * Pros
   * Most efficient model
   * Easier to aggregate stats
   * Easier to collect cluster state
 * Cons
    *   Non trivial upgrades
    *   Non trivial Authn + Authz between CP and Envoy
    * Implement & Maintain a DP Sidecar Agent / SDS Server
 
  
Relates to https://github.com/envoyproxy/gateway/issues/40 & https://github.com/envoyproxy/gateway/issues/41

**Comment by alexgervais** [comment]
A couple of questions...

- In option 3, does the CP watch for a single GatewayClass (and all of its Gateway), or does it watch for all GatewayClass?
- Adding a provisioner to option 3, how would it be different to option 2?

And some thoughts...

- Aiming for simplicity, option 1 seems the easiest to secure, operate, "get started", and has the lowest friction for common operations like upgrades (prevents version drifts between CP and DP)... although it seems like an antipattern in most Envoy Proxy deployments. The issue is scalability and fine-tuning.
- We might simplify option 2 by including the provisioner inside the CP (to manage Envoy Proxy fleet) as opposed to having a CP-provisioner.  This would better align with decisions to run a single component. https://github.com/envoyproxy/gateway/issues/60, https://github.com/envoyproxy/gateway/pull/87

**Comment by danehans** [comment]
+1 for option #2. IMHO it strikes a balance between options 1 and 3 (simplicity and flexibility).

**Comment by lizan** [comment]
Yeah agreed on 2 or 3. IMO they might not be different that much at the end, so we can start with 2 and potentially support 3 model in the future.

**Comment by arkodg** [comment]
I think 2 & 3 are great long term solutions that allow the project to handle higher scale as well as consolidate multiple config classes /`GatewayClass`.
However I'd prefer if the project started with 1
* This would allow the initial phase of the project to concentrate on UX, ease of use and features
* This would add less operational burden on early users adopting this project who might need to upgrade often since the project will likely end up releasing more often in the beginning.
* 1 will be easier to test & run adding more velocity to the project

**Comment by youngnick** [comment]
I think that 1 is not a good idea. Although it *seems* like it's easier to manage, it adds big problems later once you're in prod, in that it's much harder to do upgrades without dropping traffic than if you can upgrade the control plane and data plane independently. This is why Contour moved from a single-Pod model like 1 to the split model it has today.

I am very against a multitenanted control plane model because whatever engineering effort we save in having less things running, we will have to pay in careful security engineering to ensure that it's impossible for one data plane tenant to access another data plane tenant's data. This is particularly important in the ingress case, because the secret data often includes keypairs for whole site serving TLS certificates, or even worse, keypairs for wildcard serving certs. It is absolutely a solvable problem, but by keeping the differrent parts separate, we separate the security blast radius as well as the operational one.

In terms of authenticating between the control plane and Envoy, Contour has solved this problem in the simple case, using an on-disk SDS for hot-reload of certificates, an install process that generates one-year validity certs for both Contour and Envoy, and updates them on each upgrade. The update is seamless because the secrets change "on disk", Envoy sees the SDS change, and then control plane connection is rebuilt in less than 30 seconds. Notably, because it uses Kubernetes Secrets as intended and can handle rotations whenever you're ready, it plugs in easily to Vault or other secret-management tools - the Secret admin just has to ensure that the Secrets are updated for both Envoy and Contour around the same time.

I would also argue that 2's upgrades, although they seem harder, are actually operationally better because they give control back to the user over when various parts are upgraded. The downside is that the control plane needs to wait some period (Contour uses at least 1 release) before it can start using features added in a new Envoy version.

Although Contour's current model installs an CP and DP per Gateway at the moment, that is *not* required by the architecture or the Gateway API . The Gateway API has explicit support for merging Gateway config, in order to facilitate resource-sharing and reduce the operating cost of running the CP.

With respect to how many GatewayClasses to watch, we don't do a good job of making this clear in the Gateway API docs, but I've always considered a GatewayClass to be equivalent to a controller plus its config (that's why there's a `paramRef` in the GatewayClass spec). That is, it's a controller, plus an associated set of config. In the style of implementation we're doing where the controller runs in-cluster, having a GatewayClass == a control plane makes both designing the config for the control plane, and the control plane itself, more straightforward.

I guess I'd summarize my thoughts here by saying that:
- I prefer option 2
- It's basically the standard Operator pattern, that provisions CPs and DPs as required to satisfy Gateways, dynamically.
- This is basically what we had in mind when we built the upstream API, so it fits well with the Gateway API constructs.
- Operational complexity is present but expected in the Operator pattern. People don't expect to be managing the CP or DP directly in this pattern, they expect the Operator to handle it for you, including upgrades. The good news is that we can codify the provisioner to be able to handle upgrades and other operations (like secret rotation), and give it our best engineering towards making the operations happen with minimal user intervention.
- In terms of having to build things, we should have to build very little, because we can lift a *lot* of what Contour is doing already.

I'd definitely like to hear from @LukeShu about option 1, since that's what Emissary does today, and I remember him having some concerns about it, but I can't remember if they were distinct from what I mentioned above.

**Comment by wilsonwu** [comment]
+1 on option 2, as same as Contour, it is OK for extension in future, because it is already compatible with option 1. And option 2+3 both fine and for more modes.

**Comment by arkodg** [comment]
> I think that 1 is not a good idea. Although it _seems_ like it's easier to manage, it adds big problems later once you're in prod, in that it's much harder to do upgrades without dropping traffic than if you can upgrade the control plane and data plane independently. This is why Contour moved from a single-Pod model like 1 to the split model it has today.

@youngnick could you please elaborate on this. AFAIK with 1, you would upgrade the DP as often as you would upgrade the CP. in 2 or 3, you _could_ upgrade the CP more often than the DP (since DP upgrade would be opt in). In both cases the design for hitless upgrades based on L3 LB health checks, [draining procedure](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/draining) and outlier detection for envoy upstream, would be the same.

**Comment by danehans** [comment]
xref https://github.com/envoyproxy/gateway/issues/103

**Comment by youngnick** [comment]
> @youngnick could you please elaborate on this. AFAIK with 1, you would upgrade the DP as often as you would upgrade the CP. In 2 or 3, you _could_ upgrade the CP more often than the DP (since DP upgrade would be opt in). In both cases the design for hitless upgrades based on L3 LB health checks, [draining procedure](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/draining) and outlier detection for envoy upstream, would be the same.

It's not that people generally upgrade the things separately because they want to run different version - in Contour we pin a Contour version to only be supported with a particular Envoy version rather than a range, which I think EG should probably do too. It's the fact that changing out the CP and DP separately provides *operational* flexibility in that the change doesn't have to be atomic, and if something goes wrong with the control plane upgrade, the data plane keeps working.

It also means that the provisioner can have the option to do the upgrades in either order (control plane first or data plane first) if we need to. Again, in Contour we've found cases where upgrading the two pieces in separate relatively-atomic deploys has avoided operational problems. The tradeoff is that, although we say Contour doesn't support anything other than one version of Envoy, we need to ensure that a new version of Envoy won't break anything, which practically has meant that we've needed to go a little slower on using brand new Envoy features, waiting until one release after we required a particular Envoy version to use features that were new in that version.

I guess a better way to say all of that is maybe to say that the split model manages the operational risk better, splitting the failure domains for any individual change up.

**Comment by danehans** [comment]
> Operational complexity is present but expected in the Operator pattern. People don't expect to be managing the CP or DP directly in this pattern, they expect the Operator to handle it for you, including upgrades. The good news is that we can codify the provisioner to be able to handle upgrades and other operations (like secret rotation), and give it our best engineering towards making the operations happen with minimal user intervention.

This leads me to naming... the "Provisioner" is more than a provisioner. The "Provisioner" is more of a "Manager" or "Operator" that manages/operates the CP/DP.

Definitions:
- Provision- The action of providing or supplying something for use.
- Manage- The process of dealing with or controlling things or people.
- Operate- Control the functioning of (a machine, process, or system).

**Comment by danehans** [comment]
> in Contour we pin a Contour version to only be supported with a particular Envoy version rather than a range, which I think EG should probably do too. 

+1 @LukeShu @alexgervais I would appreciate your vote on this too.

To add to https://github.com/envoyproxy/gateway/issues/97#issuecomment-1151707468, #2 better positions EG to support canary upgrades.

I am a +1 for option #2 as the EG Kube deployment model. @arkodg due to https://github.com/envoyproxy/gateway/issues/106, can you please update the name of this issue to reflect that this issue is focused on the Kube deployment mode?

**Comment by arkodg** [comment]
started a brainstorming doc [here](https://docs.google.com/document/d/1kTFe5c8x53RCZdi2GLUa5_uGzEPwM2-na-hjBdZcEgA/edit), please feel free to share any thoughts that'll have drive the decision

**Comment by alexgervais** [comment]
All things considered, I'd go with option 1, or a hybrid of 1 and 2 where the Proxy DP doesn't run in the same process/pod as the Gateway CP.

I do not want to add operational complexity and have the DP and CP version coupling drift, or even be a concern to the administrators. We should pin a single Envoy Proxy version per Gateway release and manage the DP lifecycle in the CP. The Kubernetes constructs already mitigate the risks related to rolling out an update to the CP while ensuring uptime on the DP.

As far as the CP operator/provisioner goes, I feel this is a concern only for Kubernetes deployments, and we can do without it right now by providing a CP Helm chart. We might want to add an operator component in the future, but we don't need this complexity right now since the project's value is in the Gateway CP.

**Comment by youngnick** [comment]
After the meeting yesterday, I'd summarize the state as follows: All the maintainers present were broadly in agreement that we should go with something roughly like model #2 in https://github.com/envoyproxy/gateway/issues/97, but there are two further options about exactly how it will work, essentially the Arko proposal and the Nick proposal.

Arko and I have written a small doc about what these two proposals have in common and don't have in common, that I'll reproduce here, with @arkodg's consent.

Option A with Operator, @youngnick proposal
![Option A](https://user-images.githubusercontent.com/9346599/173962848-9f5afde7-bc70-4c1b-a562-75ca16078e75.jpg)

Option B, Operator Optional, @arkodg proposal
![Option B](https://user-images.githubusercontent.com/9346599/173962940-026427d7-9399-4196-b1d1-4598ec817c92.jpg)

## What’s Common

* The Operator watches for GatewayClass resources and provisions a unique Control Plane per GatewayClass resource. There is a 1:1 Mapping between a GatewayClass and Control Plane
* The Control Plane watches for Gateway and HTTPRoute (for now) resources
* There is a 1:1 Mapping between a GatewayClass and Data Plane (Envoy Proxy)
* All Gateway resources are merged and applied to a single DataPlane associated with the GatewayClass.
* The Control Plane translates Gateway and HTTPRoute resources into xDS resources and applies it in Envoy.

## What’s Different

* In A, the Operator provisions the Data Plane. In B, the Control Plane provisions the Data Plane
* In A, the Operator watches for GatewayClass and Gateway resources and passes the Gateway resources to watch to the Control Plane via configuration. In B, the Control Plane watches the Gateway resources.
* The A based model splits up Operator and the Control Plane into which output resources they manage, Operator manages provisioning Kubernetes resources and the Control Plane manages xDS resources. The B based model splits up Operator and the Control Plane into component hierarchy, the Operator manages the Control Plane and the Control Plane manages the Data Plane.

## Pros for A
* The design has a functional split between the Operator and the control plane, which makes their scope smaller.
  * The Operator listens for Kubernetes resources and creates Kubernetes objects in response.
  * The control plane listens for Kubernetes resources and translates it into xDS config for Envoy to pull.
* Mandating the Operator allows us to ease the management burden on the end user; they install the Operator, then just create the other Gateway API resources, and the Operator does the rest. It also ensures a consistent Envoy Gateway experience for users.
* Having an Operator also allows us to encode upgrades and other management operations into it, allowing the subject matter experts (ie, us) to solve that problem for our users.

## Cons for A
* It does require more access to the cluster; the Operator is a cluster-level service and needs quite a high level of cluster access. [youngnick] I don’t feel that the extra security requirements for the operator are that big a change, since the control plane by itself must have read access to Gateways, HTTPRoutes, Services, Endpoints, and Secrets from the entire cluster anyway for it to be able to function properly. The Operator adds write access to  Namespaces, and then inside the namespaces it creates, write access to Deployments, Services, Daemonsets, and Secrets. (We can actually ensure this by carefully setting the cluster-level permissions that we grant the Operator, and having it create Roles inside the Namespaces it creates rather than just using ClusterRoles).

## Pros for B
* B makes the Operator optional.
  * Projects such as Istio have [deprecated their Operator](https://istio.io/latest/docs/setup/install/operator/) due to security reasons.
  * Some users tend to directly use Helm and skip using the Operator to provision the Control Plane.
  * Not running the operator is one less deployment to manage.

**Comment by youngnick** [comment]
I am +1 for Option A.

**Comment by arkodg** [comment]
+1 for Option B

**Comment by youngnick** [comment]
Could I please ask all maintainers to cast binding votes on this issue before next Monday, 9am PST? That's @youngnick, @arkodg, @danehans, @skriss, @alexgervais, @LukeShu. At that time, we'll count up the existing votes and call this issue resolved.

We can also discuss at tomorrow's meeting if people have more questions.

Non-binding votes from other community members are definitely welcomed.

**Comment by danehans** [comment]
I am leaning towards Option B but I'm waiting to cast my vote until one issue is resolved.

My reasoning for leaning towards Option B:
1. Provides choice- Users can choose to use the operator or not. Some users may use in-house tooling to manage EG or simply `kubectl apply -f quickstart.yaml`. Although the operator will be used commonly for production environments, I expect most developers to not use it.
2. Aligns well with simplicity and "batteries included" project goals. A single process/container is required to run the control plane and it provisions the data plane. This is important when we're comparing Envoy Gateway/Envoy UX with Nginx.
3. Simplified troubleshooting- All the Gateway API controllers reside in the control plane, I have one place to tail logs, etc. to perform troubleshooting.
4. Provides a clear mental model- The operator is the management plane that manages the control plane. The control plane controls the data plane.
5. More secure- Supports users with very restrictive security requirements. Although the operator can be tuned to be more secure, providing namespace write access is a highly privileged operation.

My issue stems from the __What's Common__ section:
> The Operator watches for GatewayClass resources...

With Option B, all Gateway API controllers should reside in the control plane. The operator reconciles a CRD, e.g. `EnvoyGateway`, and creates the control plane (GatewayClass, Gateway, Deployment, etc). This is the well-established operator/operand [design pattern](https://www.cncf.io/wp-content/uploads/2021/07/CNCF_Operator_WhitePaper.pdf). When the control plane starts, the Gateway API controllers begin reconciling the GatewayClass/Gateway resources that were created by the operator. Since the GatewayClass controllerName matches the control plane's configured controllerName, it updates the status of the GatewayClass indicating it manages the object. It then reconciles the Gateway and creates the data plane infrastructure, e.g. Envoy Service, Envoy Deployment, RBAC, etc. When a user creates another Gateway of the same GatewayClass, the control plane will merge the config and push it to the data plane. If another control plane is desired, the user repeats the same process.

In summary, the operator reconciles a CRD to manage the control plane and the control plane reconciles Gateway API objects to manage the data plane.

**Comment by LukeShu** [comment]
I honestly don't have a strong opinion between the two.

**Comment by arkodg** [comment]
> In summary, the operator reconciles a CRD to manage the control plane and the control plane reconciles Gateway API objects to manage the data plane.
SGTM

**Comment by alexgervais** [comment]
+1 on option B, with the caveat highlighted by @danehans.

**Comment by skriss** [comment]
+1 for Option A, I prefer the functional split. Also, given that we're most likely going to end up with an `envoy-gateway-system` namespace where all the CP/DP are deployed (since they correspond to a GatewayClass which is a cluster-scoped resource), the RBAC concerns are minimized.

**Comment by lizan** [comment]
+1 for Option B, also note that A isn't really a clean functional split as the Envoy provisioning involves bootstrap configuration which is basically an xDS handling. There will be requests to customize that part as well, as we've seen that in the past in Istio.

**Comment by youngnick** [comment]
Out of our six binding votes, that puts us at:
- Option A : 2, @youngnick and @skriss 
- Option B : 2, @alexgervais and @arkodg
- Abstain : @LukeShu
- Undecided : @danehans

Looks like the decision is up to you, @danehans.
